### ISSUE TITLE
Support for `git commit --fixup` and `git rebase -i --autosquash`

### ISSUE THREAD

**Issue by AdrienMatricon** [sourcegit-scm/sourcegit#588]
The context:
- As you know, when committing a fix for the latest commit, you can run `git commit --amend` to directly modify that commit
- Committing a fix for an older commit can also be done, but in two steps
  * Passing `--fixup XXX` to the `git commit` command (where XXX is the commit's hash) creates the commit, gives it a fixup name, and marks it as a fixup commit for XXX
  * Passing `--autosquash` to `git rebase -i` automatically moves fixup commits after the relevant commits (if they're listed), an puts the "fixup" keyword in front of them instead of the default "pick"

The issue:
- `sourcegit` does not support those options
- More generally, it does not try to make it easy for the user to handle fixup commits (beyond the existence of an interactive-rebase window)

Possible solutions:
- The most direct approach would be to separately implement both options:
  * When right-clicking on a commit, add an option `Commit staged changes as a fixup to here`
  * In the interactive-rebase window, add at the top a button `Toggle autosquash` (maybe with a warning that it cancels and replaces the current interactive rebase if someone had started to edit it)
- A more integrated approach would be to allow the user to get the same result without exposing the underlying git commands
  * When right-clicking on a commit, add an option `Try to integrate the staged changes as a fixup to here (using rebase)`
    + Which runs `git commit --fixup XXX` (because it creates a commit with a proper name)
    + Then runs `git rebase -i` on the commit's parent (without autosquash to avoid side-effect)
    + Then sets the fixup commit after XXX with the "fixup" keyword in front of it
    + Then triggers the rebase as if the user has done all that manually (i.e. let them handle conflicts as usual)
  * If you choose that solution, you can reuse the logic to implement an additional feature : when right-clicking on a commit with other commits selected , add an option `Try to squash the selected commits to here (using rebase)`
    + Which runs `git rebase -i` on the commit's parent
    + Then sets the selected commits after it with the "squash" keyword in front of them
    + Then triggers the rebase as if the user has done all that manually (i.e. let them handle conflicts as usual)

**Comment by cdammanintopix** [comment]
Indeed, it would be nice to have better support for fixup commits! 🙂

**Comment by mithom** [comment]
is this issue not resolved with all the interactive rebase option changes? that would mean it can be closed.

**Comment by gotoh** [comment]
I'm interested in this feature. It seems that fork.exe can do this by specifying `Fork.RI.exe` ("RI" is "rebase interaction"?) as the editor when using `git rebase -i` to get autosquashed sequence created by git.

**Comment by love-linger** [comment]
You can also do this in ‘Interactive Rebase’ dialog of SourceGit

1. Create a new commit
2. Open Interactive Rebase dialog from the context menu of the target commit you want to run ‘git rebase -I’
3. Reorder the commits to make sure the new commit is just on top of the commit you want to fix
4. Chang the action of the new commit to  ‘Fixup’ and start rebasing

**Comment by gotoh** [comment]
As an example...

I have a repository with a commit made by "git commit -m D --fixup HEAD~~"

`fork` shows commits like this.
<img width="525" height="108" alt="Image" src="https://github.com/user-attachments/assets/06ede730-ef70-4203-9a8a-62039d625d5b" />

Then interactive rebase on `7cbf3a4` shows like this:
<img width="657" height="175" alt="Image" src="https://github.com/user-attachments/assets/9b2f91e8-87ba-4eac-9971-cf894156e625" />

The last commit is already moved, and its action is changed to "fixup"
when window is appeared.

Of course, `git` makes same list.
`git rebase -i --autosquash 7cbf3a4` makes this sequence.
```
pick 03ef181 # B
fixup e9a4f32 # fixup! B
pick d3b40eb # C

# Rebase 7cbf3a4..e9a4f32 onto 7cbf3a4 (3 commands)
#
# Commands:
...
```

This is what I want.
It seems that `fork` get this ordered sequence passed by `git.exe` by setting `forkRI.exe` as editor for `git`.

But SourceGit shows...

<img width="644" height="157" alt="Image" src="https://github.com/user-attachments/assets/ee623834-c695-41a5-8e6b-f84d94d0a709" />

I hope SourceGit aware to autosquash behavior.

**Comment by love-linger** [comment]
This is because when interactive rebase is opened in the fork, the rebase operation has actually started to run, so it can read the todo list. This is also why fork is very time-consuming when opening and closing the IR window. In SourceGit, the git rebase command will be actually executed only after clicking the start button

**Comment by love-linger** [comment]
I've pushed a new commit that reorders fixup commits when loading commits for interactive rebasing

**Comment by love-linger** [comment]
Here is the way to run `git commit --fixup` and `git rebase -i --autosquash` in `SourceGit`:

- Create a fixup commit by one of following two ways:
  * Use command line `git commit --fixup=<commit_hash>`
  * Copy the target commit's subject in `SourceGit` and create a commit with message `fixup! <COPIED_SUBJECT>`
- Select the parent of target commit (the commit you want to fixup) and choose context menu entry `Interactive Rebase`

**Comment by gotoh** [comment]
Fantastic!!
It would also be great if you could support "--squash".

BTW, "git commit --fixup=xxx" can be defined in custom action?

**Comment by love-linger** [comment]
> BTW, "git commit --fixup=xxx" can be defined in custom action?

Of cause you can. 

1. Create a custom action (scope `Commit`)

<img width="600" height="440" alt="Image" src="https://github.com/user-attachments/assets/0408add5-b947-4bad-8acf-36d96c4e0926" />

2. Stage files to fixup target commit
3. Select the commit you want to fixup, and choose the context menu entry `Custom Action > Fixup with Staged Files`

<img width="570" height="395" alt="Image" src="https://github.com/user-attachments/assets/abbb25a6-2e26-408e-a6d3-0ed44d59f9e7" />

**Comment by love-linger** [comment]
I've pushed a new commit that reorders `squash!` commits to its right position when loading commits for interactive rebasing

**Comment by gotoh** [comment]
Good!  but I found strange cases. See #2261.

**Comment by cdammanintopix** [comment]
Nice! Thank you for the development of this amazing feature!
Would it be possible to have an integrated menu (instead of a custom action) that does the fixup commit and directly launches the interactive rebase popup?

**Comment by love-linger** [comment]
Currently, there's no plan to add this as a built-in feature.

- The UI/UX appears unreasonable. The `LOCAL CHANGES` page where changes are located and the `HISTORY` page where the target commit located are two separate pages. Users cannot complete or verify the correctness of their operations in a single page.
- If a new context menu entry, such as `Fixup with Staged File`, is added for each commit, it will coexist with `Interactive Rebase > Fixup`, which may confuse users and make them unable to understand the difference between the two.
